keps/sig-scheduling/5710-workload-aware-preemption/README.md

```diff
@@ -0,0 +1,1064 @@
+    // should be used when scheduling this workload. "system-node-critical" and
+	// "system-cluster-critical" are two special keywords which indicate the
+	// highest priorities with the former being the highest priority. Any other
+	// name must be defined by creating a PriorityClass object with that name.
+	// If not specified, the priority will be default or zero if there is no
+	// default.
+    //
+    // This field is immutable.
+    PriorityClassName *string
+
+    // PreemptionPriorityClassName, if specified, indicates the workload's
+    // priority that should be used when attempting to preempt this workload.
+    // If not specified, it will default to PriorityClassName.
+    //
+    // This field is mutable.
+    PreemptionPriorityClassName *string
+}
+```
+
+If that appears not being enough, we will similarly extend the `PodGroup` API in a follow up.
+In such case, the priority of a pod would be the priority of a smallest unit in the `Workload`
+object (Workload, PodGroup, PodSubGroup, ...) corresponding to this pod.
+
+
+There is one direct implication of the above - the `pod.Spec.PriorityClassName` and `pod.Spec.Priority`
+may no longer reflect the actual pod priority. This can be misleading to users.
+
+```
+<<[UNRESOLVED priority divergence]>>
+There are several options we can approach it (from least to most invasive):
+- Explain via documentation
+- Validating that if a pod is referencing a workload, `pod.Spec.PriorityClassName` equals
+  `workload.Spec.PriorityClassName`. However, `Workload` object potentially may not exist
+  yet on pod creation.
+- Making `pod.Spec.PriorityClassName` and `pod.Spec.Priority` mutable fields and having a
+  controller responsible for reconciling these. However, that doesn't fully address the
+  problems as divergence between the pod and PodTemplate in true workload object could also
+  be misleading.
+
+The validation option seems like the best option, if we can address the problem of not-yet
+existing `Workload` object (reversed validation?).
+<<[/UNRESOLVED]>>
+```
+
+The similar argument holds for preemption priority, but we argue that its mutable nature
+makes it infeasible for reconciling this information back to pod for scalability reasons
+(we can absolutely handle frequent updates to `Workload.Spec.PreemptionPriorityClassName`
+but we can't handle updating potentially hundreds of thousands of pods within that workload
+that frequently). In this case, we limit ourselves to documentation.
+
+```
+<<[UNRESOLVED priority status]>>
+We should introduce/describe `workload.status` to reflect:
+- actual priority (a number) that is inferred from the priority class
+- express the preemption priority that is currently used by scheduler to reflect that
+  it aknowledged the change
+<<[/UNRESOLVED]>>
+```
+
+### Preemption algorithm
+
+We start with describing at the high-level how existing pod-level preemption algorithm works.
+Below, we will show how to generalize it to workloads.
+
```

Comment:
Currently preemption is not even considered unless Filter returns no results.

- Maybe mention that here.
- There are scheduling algorithms that we might someday want to run which preempt even when there is a feasible placement without preeemption.  IIUC, you are putting that out of scope for this KEP, so maybe state that.